core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,570 @@
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    if (cmk->keybloblen == 0)
+    {
+        cmk->keybloblen = real_blob_len;
+        return SGX_SUCCESS;
+    }
+
+    uint8_t *key = (uint8_t *)malloc(keysize);
+    if (key == NULL)
+    {
+        return SGX_ERROR_OUT_OF_MEMORY;
+    }
+    ret = sgx_read_rand(key, keysize);
+    if (ret != SGX_SUCCESS)
+    {
+        free(key);
+        return ret;
+    }
+    ret = ehsm_create_keyblob(key,
+                              keysize,
+                              NULL,
+                              0,
+                              (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+
+    memset_s(key, keysize, 0, keysize);
+
+    free(key);
+    return ret;
+}
+
+sgx_status_t ehsm_create_rsa_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk == NULL)
+    {
+        return ret;
+    }
+
+    if (cmk->keybloblen == 0)
+    {
+        cmk->keybloblen = PEM_BUFSIZE * 5;
+        return SGX_SUCCESS;
+    }
+
+    EVP_PKEY_CTX *pkey_ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    BIO *bio = NULL;
+    uint8_t *pem_keypair = NULL;
+    uint32_t key_len = 0;
+
+    pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+    if (pkey_ctx == NULL)
+    {
+        goto out;
+    }
+
+    if (EVP_PKEY_keygen_init(pkey_ctx) <= 0)
+    {
+        goto out;
+    }
+
+    switch (cmk->metadata.keyspec)
+    {
+    case EH_RSA_2048:
+        key_len = RSA_2048_KEY_BITS;
+        break;
+    case EH_RSA_3072:
+        key_len = RSA_3072_KEY_BITS;
+        break;
+    case EH_RSA_4096:
+        key_len = RSA_4096_KEY_BITS;
+        break;
+    default:
+        break;
+    }
+
+    if (EVP_PKEY_CTX_set_rsa_keygen_bits(pkey_ctx, key_len) <= 0)
+    {
+        goto out;
+    }
+
+    if (EVP_PKEY_keygen(pkey_ctx, &pkey) <= 0)
+    {
+        goto out;
+    }
+
+    bio = BIO_new(BIO_s_mem());
+    if (bio == NULL)
+    {
+        goto out;
+    }
+
+    if (!PEM_write_bio_PUBKEY(bio, pkey))
+    {
+        goto out;
+    }
+
+    if (!PEM_write_bio_PrivateKey(bio, pkey, NULL, NULL, 0, NULL, NULL))
+    {
+        goto out;
+    }
+
+    key_len = BIO_pending(bio);
+    if (key_len <= 0)
+    {
+        goto out;
+    }
+
+    pem_keypair = (uint8_t *)malloc(key_len + 1);
+    if (pem_keypair == NULL)
+    {
+        goto out;
+    }
+
+    if (BIO_read(bio, pem_keypair, key_len) < 0)
+    {
+        goto out;
+    }
+    pem_keypair[key_len] = '\0';
+
+    ret = ehsm_create_keyblob(pem_keypair, key_len, NULL, 0, (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+
+    if (ret != SGX_SUCCESS)
+    {
+        goto out;
+    }
+out:
+    EVP_PKEY_CTX_free(pkey_ctx);
+    EVP_PKEY_free(pkey);
+    BIO_free(bio);
+    SAFE_FREE(pem_keypair);
+    return ret;
+}
+
+sgx_status_t ehsm_create_ec_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk == NULL)
+    {
+        return ret;
+    }
+
+    if (cmk->keybloblen == 0)
+    {
+        cmk->keybloblen = PEM_BUFSIZE;
```

Comment:
same here, why not use ehsm_calc_keyblob_len?

Reply:
it fixed